### ISSUE TITLE
Vendor neutral to configure a Tracer

### ISSUE THREAD

**Issue by fmhwong** [opentracing-contrib/java-tracerresolver#35]
Tracer resolver is supposed to be a vendor neutral way to obtain a Tracer implementation.  If I have multiple vendor implementations available, it can resolve with the highest priority implementation.  However, there isn't a standard way to configure the Tracer in a vendor neutral way.

For example, Jaeger requires to set the Jaeger specific environment variables or system variables before creating a Tracer.
 
```
Tracer tracer = TracerResolver.resolveTracer();
```

It seems the configuration cannot be modified after a Jaeger tracer has be created.  If there are multiple implementations available, how do I know which implementation I will get and set the required vendor specific configuration?

**Comment by sjoerdtalsma** [comment]
Think of it as the [slf4j LoggerFactory](https://github.com/qos-ch/slf4j/blob/master/slf4j-api/src/main/java/org/slf4j/LoggerFactory.java). From your code, you just use the factory or resolver to _obtain_ the implementation, while the configuration is done elsewhere.

The TracerResolver provides such a way, individual tracers have their own configuration rules that must still be followed.

When building your application, you don't have to use the TracerResolver. You can also use a dependency injection framework such as guice, dagger, spring, JEE, or use the `GlobalTracer` from opentracing-util.

**Comment by fmhwong** [comment]
By default, Jaeger reads it configuration from environment variables including service name.  It is probably okay to set agent host and port from the environment variables as they are static.  If I want to have more than one Tracer, it will be nice if I can set the service name in the `resolveTracer()` method.  

For example,  `Tracer tracer = TracerResolver.resolveTracer(serviceName);`

**Comment by sjoerdtalsma** [comment]
As far as I know, service name is configuration at tracer level, specific for some tracers*.
As the resolver is vendor neutral, there is no way to specify this to _any_ tracer, unless the service name concept were included in the opentracing api itself.

*) I think at least zipkin/brave and Jaeger support this when configuring a new tracer.

**Comment by fmhwong** [comment]
Tracers usually require configurations.  How about adding a new method and passing the configuration values into a java.util.Properties object as a parameter?

```
public static Tracer resolveTracer(java.util.Properties properties)
```

Implementers can decide what to do with the Properties values or simply ignore them if they don't support it.

**Comment by sjoerdtalsma** [comment]
Quite frankly that sounds like kind of a random mechanism, choosing for properties. Why not use system properties or environment variables for the same purpose?
Why not explicitly initialise the tracer yourself, as you have to know the implementation upfront anyway to be sure you provide the correct properties.

Unless you manage to unify all tracer providers on their initialisation options, I don't see how adding properties to the resolver is going to help you. Could you create an example of how you need the `resolveTracer(Properties)` method for your usecase?

**Comment by fmhwong** [comment]
To get a Tracer object from Jaeger API for service name "serviceA":

```
Configuration configuration = new Configuration("serviceA");
Tracer tracer = configuration.getTracer();
```

Currently, there is no way to do this from TracerResolver class.  What I am proposing is to have something like this.

```
Properties properties = new Properties();
properties.setProperty("JAEGER_SERVICE_NAME", "serviceA");
Tracer tracer = TracerResolver.resolveTracer(properties);
```

The service name can also be set using `System.setProperty("JAEGER_SERVICE_NAME", "serviceA");`.  This may require synchronization if it runs on a multi-thread environment. Comments?

**Comment by sjoerdtalsma** [comment]
I'm sorry, I still don't get your use-case.  
If you are configuring Jaeger from code anyway, why would you _want_ to use the TracerResolver instead of just creating a new Jaeger tracer?

**Comment by fmhwong** [comment]
I would like to separate the Open Tracing code from the implementation.  On the server, I would like to use pure Open Tracing APIs and I don't want to include any vendor specific implementation jar files.  The person who deploy the server can choose a list of compatible implementations and provide the jar files for that implementation.

**Comment by sjoerdtalsma** [comment]
How is providing jaeger-specific properties separating from the implementation?

**Comment by fmhwong** [comment]
We don't need to include the Jaeger jars into the server binaries and we don't need to worry about upgrading it on the server when there is a newer version.  It is similar to JDBC drivers, server binaries do not include them but you can configure database specific settings in your server configuration.

**Comment by xificurC** [comment]
This is a bit old, but for those who stumble into this: I think this is an issue that needs to be raised for jaeger, not this project. `resolveTracer()` is all one needs. If you have a case where you're creating multiple jaeger configurations programmatically then yes, you're in trouble because the jaeger implementation calls `Configuration.fromEnv().getTracer()` without giving you a chance to somehow pick something else. Setting up jaeger from system properties or environment variables is recommended by jaeger, if that isn't sufficient for you you'll need to raise that issue with them.

**Comment by fmhwong** [comment]
`tracerresolver` has only one implementation which is Jaeger.  You know you are going to get a `JaegerTracer` instance so you set the environment variables for Jaeger before you call `resolveTracer()`.  If there is more than one implementation, this design won't work.

**Comment by xificurC** [comment]
Exactly, because `resolveTracer()` is about returning 1 `Tracer`, not picking from a number of them. Your proposal needs a better name, because the current implementation allows the configuration to be vendor neutral, it just expects you to configure 1 tracer.
